src/dds/rs-dds-device-proxy.cpp

```diff
@@ -534,20 +539,115 @@ std::vector< uint8_t > dds_device_proxy::build_command( uint32_t opcode,
-                                                     { "build-command", true } } );
-    rsutils::json reply;
+    json control = rsutils::json::object( { { realdds::topics::control::key::id, realdds::topics::control::hwm::id },
+                                            { realdds::topics::control::hwm::key::data, hexdata },
+                                            { realdds::topics::control::hwm::key::opcode, opcode },
+                                            { realdds::topics::control::hwm::key::param1, param1 },
+                                            { realdds::topics::control::hwm::key::param2, param2 },
+                                            { realdds::topics::control::hwm::key::param3, param3 },
+                                            { realdds::topics::control::hwm::key::param4, param4 },
+                                            { realdds::topics::control::hwm::key::build_command, true } } );
+    json reply;
     _dds_dev->send_control( control, &reply );
-    if( ! reply.nested( "data" ).get_ex( hexdata ) )
+    if( ! reply.nested( realdds::topics::reply::hwm::key::data ).get_ex( hexdata ) )
         throw std::runtime_error( "Failed HWM: missing 'data' in reply" );
     return hexdata.detach();
 }
 
 
+bool dds_device_proxy::check_fw_compatibility( const std::vector< uint8_t > & image ) const
+{
+    try
+    {
+        // Start DFU
+        json reply;
+        _dds_dev->send_control(
+            json::object( { { realdds::topics::control::key::id, realdds::topics::control::dfu_start::id } } ),
+            &reply );
+
+        // Set up a reply handler that will get the "dfu-ready" message
+        std::mutex mutex;
+        std::condition_variable cv;
+        json dfu_ready;
+        auto subscription = _dds_dev->on_notification(
+            [&]( std::string const & id, json const & notification )
+            {
+                if( id != realdds::topics::notification::dfu_ready::id )
+                    return;
+                std::unique_lock< std::mutex > lock( mutex );
+                dfu_ready = notification;
+                cv.notify_all();
+            } );
+
+        // Upload the image; this will check its compatibility
+        auto topic = realdds::topics::blob_msg::create_topic( _dds_dev->participant(),
+                                                              _dds_dev->device_info().topic_root()
+                                                                  + realdds::topics::DFU_TOPIC_NAME );
+        auto writer = std::make_shared< realdds::dds_topic_writer >( topic );
+        writer->run( realdds::dds_topic_writer::qos( eprosima::fastdds::dds::RELIABLE_RELIABILITY_QOS ) );
+        if( ! writer->wait_for_readers( { 3, 0 } ) )
+            throw std::runtime_error( "timeout waiting for DFU subscriber" );
+        auto blob = realdds::topics::blob_msg( std::vector< uint8_t >( image ) );
+        blob.write_to( *writer );
+        if( ! writer->wait_for_acks( { 3, 0 } ) )
```

Comment:
We should take timeouts from settings `reply-timeout-ms`. Can default to 3 seconds (for device-impl we default 2 seconds).

Reply:
This is not a regular control. We're sending a big blob which may have retries (reliable), so I don't think it should be the regular reply timeout. But I agree, it should be configurable and we can talk about the actual number.

I'll make all the configuration changes in the next PR.